### PR TITLE
Allow access to different nutpie backends via pip-style syntax

### LINKED ISSUE #7497: Expose all `nutpie` compile backends through `pm.sample`
### Description

Nutpie currently has two compile modes, numba and JAX, with a 3rd pytorch backend on the way. It would be nice if we could easily access these via `pm.sample`.

Proposal 1: Allow `nutpie.compile_pymc` kwargs in `nuts_sampler_kwargs`
- Pros: It's easy, since there are only two such arguments: `backend` and `gradient-backend`. We just check for and pop them before forwarding all other arguments to `nutpie.sample`. 
- Cons: It might be see as "unexpected" behavior, since some keywords are going to one function, and some to another. Also, the `nuts_sampler_kwargs` argument isn't very beautiful in the first place

Proposal 2: pip-style optional arguments, like `nuts_sampler="nutpie[jax]"` and `nuts_sampler="nutpie[numba]`"
- Pros: It's quite pretty!
- Cons: technically you can pick both the forward and backward compile mode, so if a user wanted that, she'd still have to import `nutpie` and do it manually. Maybe that's enough of a corner case that it's ok? Also it's a different API to other samplers (although blackjax could benefit from something similar to ask for the many different options over there -- but that's beyond the scope here).

Proposal 3: Add a new `compile_kwargs` argument to `pm.sample`
- Pros: It's very clear. It could be used to forward kwargs to pytensor as well, which is a nice side bonus. 
- Cons: It's another argument to an already bloated `pm.sample` function

### PR DESCRIPTION
**Pull request by jessegrabowski**
## Description

Adds a pip-style syntax to the `nuts_sampler` argument that allows access to alternative compile backends, when relevant. This lets you get the nutpie jax backend by setting `nuts_sampler='nutpie[jax]'`. For backwards compatibility, `nuts_sampler='nutpie'` is equivalent to `nuts_sampler='nutpie[numba]'`.

The current PR only deals with nutpie, but we could easily extend this to include the default PyMC sampler, to compile to JAX, numba, or pytorch directly, without going through nutpie. I'm willing to do that extension in this PR if it is deemed worthwhile..

## Related Issue

- [x] Closes #7497 
- [ ] Related to #

## Checklist

- [x] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [x] Included tests that prove the fix is effective or that the new feature works
- [x] Added necessary documentation (docstrings and/or example notebooks)
- [x] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

## Type of change

- [x] New feature / enhancement
- [ ] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ ] Other (please specify):

----
📚 Documentation preview 📚: https://pymc--7498.org.readthedocs.build/en/7498/

### CODE REVIEW

**Comment by jessegrabowski** [comment]
I created some global variables to track available backends in `nutpie`, anticipating that we will eventually expand the list to include pytorch. It would be nice if these globals were defined in the nutpie library itself and we just imported them here.

**Comment by ricardoV94** [comment]
> The current PR only deals with nutpie, but we could easily extend this to include the default PyMC sampler, to compile to JAX, numba, or pytorch directly

I'm not sure how the other backends (specially JAX) behave with multiprocessing tbh :O Otherwise the idea sounds cool. Perhaps @aseyboldt can weigh in as he has a better picture of how we handle multiprocessing in pm.sample

**Comment by jessegrabowski** [comment]
Should we go work on a `pmap` for pytensor?

**Comment by ricardoV94** [comment]
> Should we go work on a `pmap` for pytensor?

The PyMC codebase would still need to know about it and work around differently for JAX

**Comment by ricardoV94** [comment]
With numba backend we could actually do it with threads (with nogil)? Maybe worth opening an issue to investigate different backends for pymc samplers.

Also this shouldn't have to be nuts specific so for that a `backend` argument may make more sense. Something like this was also raised for VI recently IIRC

**Comment by jessegrabowski** [comment]
Should probably just be `compile_kwargs` like every other pymc function, but I liked my cute syntax.

**Comment by ricardoV94** [comment]
> Should probably just be `compile_kwargs` like every other pymc function, but I liked my cute syntax.

Not if we need to change how the samplers are orchestrated

**Comment by ricardoV94** [comment]
But compile kwargs already works anyway

**Comment by jessegrabowski** [comment]
> But compile kwargs already works anyway

What do you mean, `pm.sample` already takes `compile_kwargs`? It's undocumented if so.

**Comment by ricardoV94** [comment]
> > But compile kwargs already works anyway
> 
> What do you mean, `pm.sample` already takes `compile_kwargs`? It's undocumented if so.

It doesn't? Maybe I've played with global mode then

File: pymc/sampling/mcmc.py [hunk 1]
```diff
@@ -304,7 +308,23 @@ def _sample_external_nuts(
                 "`var_names` are currently ignored by the nutpie sampler",
                 UserWarning,
             )
-        compiled_model = nutpie.compile_pymc_model(model)
+
+        def extract_backend(string: str) -> NutpieBackend:
+            match = re.search(r"(?<=\[)[^\]]+(?=\])", string)
+            if match is None:
+                return NUTPIE_DEFAULT_BACKEND
```
Comment (lucianopaz) [inline pymc/sampling/mcmc.py]: You could also get a `None` match if the string is misformatted. For example, `nutpie[jax` would return a `None` match. I suggest that you test exact equality to set the default option, and if you get `None` then raise a `ValueError`.
```suggestion
            if string == "nutpie":
                return NUTPIE_DEFAULT_BACKEND
            elif match is None:
                raise ValueError(f"Could not parse nutpie backend. Found {string!r}")
```

File: pymc/sampling/mcmc.py [hunk 2]
```diff
@@ -304,7 +308,23 @@ def _sample_external_nuts(
...
+            result = cast(NutpieBackend, match.group(0))
+            if result not in NUTPIE_BACKENDS:
+                last_option = f"{NUTPIE_BACKENDS[-1]}"
+                expected = (
+                    ", ".join([f'"{x}"' for x in NUTPIE_BACKENDS[:-1]]) + f' or "{last_option}"'
+                )
+                raise ValueError(f'Expected one of {expected}; found "{result}"')
```
Comment (lucianopaz) [inline pymc/sampling/mcmc.py]:
```suggestion
                raise ValueError(
                    'Could not parse nutpie backend. Expected one of {expected}; found "{result}"'
                )
```

File: tests/sampling/test_mcmc_external.py [hunk 3]
```diff
@@ -74,7 +49,84 @@ def test_external_nuts_sampler(recwarn, nuts_sampler):
...
+        x = Normal("x", 100, 5)
+        y = Data("y", [1, 2, 3, 4])
+        Data("z", [100, 190, 310, 405])
+
+        Normal("L", mu=x, sigma=0.1, observed=y)
+
+    return m
+
+
+@pytest.mark.parametrize("nuts_sampler", ["pymc", "nutpie", "blackjax", "numpyro"])
```
Comment (lucianopaz) [inline tests/sampling/test_mcmc_external.py]:
```suggestion
@pytest.mark.parametrize(
    "nuts_sampler",
    ["pymc", "nutpie", "nutpie[jax]", "blackjax", "numpyro"],
)
```

File: tests/sampling/test_mcmc_external.py [hunk 4]
```diff
@@ -74,7 +49,84 @@ def test_external_nuts_sampler(recwarn, nuts_sampler):
...
+            for warn in recwarn
+            if warn.category not in (FutureWarning, DeprecationWarning, RuntimeWarning)
+        }
+
+        check_external_sampler_output(warns, idata1, idata2, sample_kwargs)
+
+
+def test_invalid_nutpie_backend_raises(pymc_model):
+    pytest.importorskip("nutpie")
+    with pytest.raises(ValueError, match='Expected one of "numba" or "jax"; found "invalid"'):
```
Comment (lucianopaz) [inline tests/sampling/test_mcmc_external.py]:
```suggestion
    with pytest.raises(
        ValueError,
        match='Could not parse nutpie backend. Expected one of "numba" or "jax"; found "invalid"',
    ):
```

File: tests/sampling/test_mcmc_external.py [hunk 5]
```diff
@@ -74,7 +49,84 @@ def test_external_nuts_sampler(recwarn, nuts_sampler):
...
+        with pymc_model:
+            sample(nuts_sampler="nutpie[invalid]", random_seed=123, chains=2, tune=500, draws=500)
+
```
Comment (lucianopaz) [inline tests/sampling/test_mcmc_external.py]:
```suggestion

    with pytest.raises(ValueError, match="Could not parse nutpie backend. Found 'nutpie[bad'"):
        with pymc_model:
            sample(nuts_sampler="nutpie[bad", random_seed=123, chains=2, tune=500, draws=500)
```

**Review by lucianopaz — CHANGES_REQUESTED** [review]
@jessegrabowski, this looks very nice. I left a few suggestions though

**Comment by twiecki** [comment]
I posted this PR with comments into GPT o1-mini:
<details>
---

#### **Summary**

This pull request introduces a pip-style syntax for specifying different Nutpie backends within the `nuts_sampler` argument of PyMC's `pm.sample` function. Users can now select a specific backend by using a syntax like `nuts_sampler='nutpie[jax]'`, with `nuts_sampler='nutpie'` defaulting to `nuts_sampler='nutpie[numba]'` for backward compatibility. This enhancement aims to provide greater flexibility and performance optimization by allowing users to choose the most suitable backend for their computational needs.

**Key Changes:**

1. **Code Enhancements:**
   - **Backend Parsing:** Introduces a helper function `_extract_backend` to parse and extract backend specifications from the `nuts_sampler` string.
   - **Type Definitions:** Defines new type aliases for better type safety and clarity.
   - **Sampler Integration:** Updates sampling functions to handle the new backend specifications seamlessly.
   - **Error Handling:** Improves error messages to provide clearer guidance when invalid backend specifications are provided.

2. **Testing Improvements:**
   - **Expanded Test Coverage:** Adds new test cases to cover various backend specifications, including valid and invalid scenarios.
   - **Refactored Tests:** Utilizes fixtures and helper functions to reduce redundancy and enhance test maintainability.

3. **Documentation Updates:**
   - **Docstrings:** Enhances existing docstrings to reflect the new backend syntax and usage examples.
   - **ReadTheDocs:** Provides a preview link to the updated documentation for further review.

4. **GitHub Discussions:**
   - **Collaborative Feedback:** Addresses feedback from reviewers, refining error handling and test cases based on suggestions.
   - **Future Considerations:** Mentions potential future expansions, such as supporting additional backends like PyTorch.

---

#### **Detailed Review**

##### **1. Code Changes**

**a. `pymc/sampling/mcmc.py`**

- **Type Aliases:**
  ```python
  ExternalNutsSampler = Literal["nutpie", "numpyro", "blackjax"]
  NutsSampler = Literal["pymc"] | ExternalNutsSampler
  NutpieBackend = Literal["numba", "jax"]
  ```
  - **Purpose:** Enhances type safety by clearly defining allowed values for samplers and Nutpie backends.
  - **Suggestion:** Consider placing these type aliases in a separate module if they are to be reused across different parts of the codebase.

- **Global Variables:**
  ```python
  NUTPIE_BACKENDS = get_args(NutpieBackend)
  NUTPIE_DEFAULT_BACKEND = cast(NutpieBackend, "numba")
  ```
  - **Purpose:** Manages available backends and sets the default backend to `numba`.
  - **Suggestion:** If the list of backends is expected to grow, consider loading them dynamically or configuring them via external settings.

- **Backend Extraction Function:**
  ```python
  def extract_backend(string: str) -> NutpieBackend:
      match = re.search(r"(?<=\[)[^\]]+(?=\])", string)
      if match is None:
          return NUTPIE_DEFAULT_BACKEND
      result = cast(NutpieBackend, match.group(0))
      if result not in NUTPIE_BACKENDS:
          last_option = f"{NUTPIE_BACKENDS[-1]}"
          expected = (
              ", ".join([f'"{x}"' for x in NUTPIE_BACKENDS[:-1]]) + f' or "{last_option}"'
          )
          raise ValueError(f'Expected one of {expected}; found "{result}"')
      return result
  ```
  - **Purpose:** Parses the backend from the `nuts_sampler` string and validates it.
  - **Strengths:** 
    - Uses regex to extract backend specifications.
    - Provides clear error messages when invalid backends are specified.
  - **Suggestions:**
    - **Regex Robustness:** Ensure the regex accurately captures backends in all valid formats and handles edge cases (e.g., multiple brackets, unexpected characters).
    - **Error Messaging:** The error message could be more descriptive by including the entire input string to aid debugging.

- **Sampler Integration:**
  ```python
  if sampler.startswith("nutpie"):
      backend = extract_backend(sampler)
      compiled_model = nutpie.compile_pymc_model(model, backend=backend)
  ```
  - **Purpose:** Integrates the backend specification into the Nutpie sampler.
  - **Suggestions:**
    - **Namespace Consistency:** Ensure that all references to samplers and backends maintain consistent naming conventions.
    - **Extensibility:** If future backends are added, verify that the integration logic remains scalable.

**b. `pymc/sampling/jax.py`**

- **Type Alias Update:**
  ```python
  JaxNutsSampler = Literal["numpyro", "blackjax"]
  ```
  - **Purpose:** Defines allowed JAX-based samplers.
  - **Suggestion:** Similar to `ExternalNutsSampler`, consider centralizing type aliases for better maintainability.

- **Function Signature Update:**
  ```python
  def sample_jax_nuts(
      ...
      nuts_sampler: JaxNutsSampler,
      ...
  ) -> az.InferenceData:
  ```
  - **Purpose:** Enforces that only `numpyro` and `blackjax` are accepted as JAX-based samplers.
  - **Suggestion:** Ensure that all downstream functions and users are aware of this restriction to prevent runtime errors.

**c. `tests/sampling/test_mcmc_external.py`**

- **Test Refactoring:**
  ```python
  @pytest.fixture
  def pymc_model():
      with Model() as m:
          x = Normal("x", 100, 5)
          y = Data("y", [1, 2, 3, 4])
          Data("z", [100, 190, 310, 405])
  
          Normal("L", mu=x, sigma=0.1, observed=y)
  
      return m
  ```
  - **Purpose:** Defines a reusable model fixture for tests.
  - **Strengths:** Promotes DRY (Don't Repeat Yourself) principles by avoiding redundancy in test definitions.

- **New Test Cases:**
  - **Backend Options Test:**
    ```python
    @pytest.mark.parametrize("backend", ["numba", "jax"], ids=["numba", "jax"])
    def test_numba_backend_options(pymc_model, recwarn, backend):
        ...
    ```
    - **Purpose:** Validates that specifying different Nutpie backends works as expected.
  
  - **Invalid Backend Test:**
    ```python
    def test_invalid_nutpie_backend_raises(pymc_model):
        pytest.importorskip("nutpie")
        with pytest.raises(ValueError, match='Could not parse nutpie backend. Expected one of "numba" or "jax"; found "invalid"'):
            with pymc_model:
                sample(nuts_sampler="nutpie[invalid]", ...)
    ```
    - **Purpose:** Ensures that invalid backend specifications raise appropriate errors.
    - **Strengths:** Provides clear assertions that verify error handling mechanisms.

- **Test Helper Function:**
  ```python
  def check_external_sampler_output(warns, idata1, idata2, sample_kwargs):
      ...
  ```
  - **Purpose:** Abstracts common test assertions to enhance readability and maintainability.
  - **Suggestion:** Ensure that this function covers all necessary assertions to comprehensively validate sampler outputs.

**d. Documentation Updates**

- **Docstring Enhancements:**
  ```python
  nuts_sampler : str, default "pymc"
      Which NUTS implementation to run. One of ["pymc", "nutpie", "blackjax", "numpyro"]. In addition, the compilation
      backend for the chosen sampler can be set using square brackets, if available. For example, "nutpie[jax]" will
      use the JAX backend for the nutpie sampler. Currently, "nutpie[jax]" and "nutpie[numba]" are allowed.
  ```
  - **Purpose:** Clarifies the usage of the `nuts_sampler` argument with examples.
  - **Strengths:** Provides users with clear guidance on how to specify different backends.

- **Documentation Preview:**
  - **Link:** [ReadTheDocs Preview](https://pymc--7498.org.readthedocs.build/en/7498/)
  - **Suggestion:** Verify that all new features are accurately reflected in the documentation, including edge cases and error scenarios.

---

##### **2. Testing Enhancements**

- **Increased Coverage:**
  - **Codecov Report:** Indicates a patch coverage of 47.83% with 12 lines missing.
  - **Action Required:** Ensure that all new code paths, especially those involving error handling and backend parsing, are adequately tested to improve coverage.

- **Comprehensive Test Cases:**
  - **Valid Backend Specifications:** Tests cover both `numba` and `jax` backends, ensuring they are correctly parsed and utilized.
  - **Invalid Backend Specifications:** Tests verify that improper backend inputs are gracefully handled with meaningful error messages.
  - **Edge Cases:** Tests include scenarios like missing brackets or unsupported backend names to ensure robustness.

- **Fixture Utilization:**
  - **`pymc_model` Fixture:** Enhances test efficiency by providing a standardized model setup for multiple tests.
  - **Suggestion:** Consider adding more fixtures if additional setup is required for future tests, promoting scalability.

---

##### **3. GitHub Discussion & Iterative Improvements**

- **Collaborative Feedback:**
  - **Error Handling Enhancements:** Reviewer @lucianopaz suggested refining error messages and handling misformatted backend specifications, which were addressed by the author.
  - **Test Case Refinements:** Suggestions to expand parameterizations and improve test assertions were implemented, enhancing test robustness.

- **Future Considerations:**
  - **Additional Backends:** The author mentioned the potential to extend support beyond Nutpie, such as integrating with PyTorch, which could be explored in future PRs.
  - **Multiprocessing Compatibility:** Discussions around how different backends behave with multiprocessing indicate areas for further investigation to ensure seamless integration.

- **Commit History:**
  - **Clarity and Relevance:** Commit messages are descriptive and align with the logical progression of changes, facilitating easier codebase navigation and history tracking.
  - **Reversion Handling:** The author promptly addressed issues raised during code review, demonstrating responsiveness and a commitment to quality.

---

#### **Conclusion**

This pull request effectively enhances PyMC's sampling functionality by introducing a flexible and user-friendly way to specify different Nutpie backends. The implementation is thoughtfully designed, with clear type definitions, robust error handling, and comprehensive testing to ensure reliability and maintainability.

**Recommendations Before Merging:**

1. **Improve Regex Parsing:**
   - **Robustness:** Ensure that the regex used in the `extract_backend` function accurately captures backend specifications and handles all valid and invalid formats.
   - **Edge Case Handling:** Consider additional tests for edge cases, such as multiple backend specifications or unexpected characters.

2. **Enhance Test Coverage:**
   - **Complete Coverage:** Address the missing lines reported by Codecov by adding tests that cover all new code paths, especially those involving error scenarios and backend extraction.
   - **Performance Tests:** If applicable, include tests that evaluate the performance impact of different backends to guide users in selecting the most suitable option.

3. **Documentation Completeness:**
   - **Usage Examples:** Incorporate detailed usage examples in the documentation to demonstrate how to specify different backends effectively.
   - **Error Messages:** Document common error messages and their resolutions to assist users in troubleshooting.

4. **Future Backend Support:**
   - **Scalability:** Plan for the potential addition of more backends by ensuring that the current implementation is easily extensible.
   - **Modular Design:** Consider refactoring the backend management logic into separate modules or classes to enhance modularity and separation of concerns.

5. **Code Cleanup:**
   - **Unused Imports:** Review and remove any unused imports or variables to maintain code cleanliness.
   - **Consistency:** Ensure that naming conventions and coding styles are consistent across the new and existing code.

Overall, this PR represents a significant improvement to PyMC's flexibility and user experience, enabling more tailored and optimized sampling strategies. With the recommended refinements, it is well-positioned for successful integration into the main codebase.

---

**Approved with minor recommendations.**
</details>

**Comment by jessegrabowski** [comment]
Close this as stale now that #7535 is merged? Or is there interest in this syntax still.

**Comment by ricardoV94** [comment]
I think this syntax is a bit more ergonomic

**Comment by aseyboldt** [comment]
I like the syntax :-)
But maybe we should go with jax by default? I know it is a change, but I think right now this is much more robust.

**Comment by jessegrabowski** [comment]
Thoughts on the best way to ask for the gradient backend in this syntax? Or `nutpie[jax]` should just always set both, and consider the mixed case to be "advanced"?

**Comment by ricardoV94** [comment]
> I like the syntax :-) But maybe we should go with jax by default? I know it is a change, but I think right now this is much more robust.

That will break people's existing code if they don't have jax installed

**Comment by AlexAndorra** [comment]
This looks super helpful @jessegrabowski !!

**Comment by ricardoV94** [comment]
I know there was some back and forth here but I think the `backend=mode` kwarg would be nicer. So you can do `pm.sample(backend="numba")` as well as `pm.sample(nuts_sampler="nutpie", backend="numba")`, and `pm.sample_posterior_predictive(backend="numba")`

That is, the backend is somewhat orthogonal to the sampler, be it pymc or not?

Basically a less verbose shortcut for the most common use of `compile_kwargs` (which can still exist for more fine tweaks)

**Comment by jessegrabowski** [comment]
Yeah I think I agree. I've cooled on all the clever string stuff recently. I think I'll close this PR and open a fresh one with the backend argument.
